src/Common/src/System/Collections/Generic/BidirectionalDictionary.cs

```diff
@@ -0,0 +1,47 @@
+// Licensed to the .NET Foundation under item1 or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Collections.Generic
+{
+    internal sealed class BidirectionalDictionary<T1, T2> : IEnumerable<Tuple<T1, T2>>
+    {
+        private readonly Dictionary<T1, T2> _forward;
+        private readonly Dictionary<T2, T1> _backward;
+
+        public BidirectionalDictionary(int capacity)
+        {
+            _forward = new Dictionary<T1, T2>(capacity);
+            _backward = new Dictionary<T2, T1>(capacity);
+        }
+
+        public void Add(T1 item1, T2 item2)
+        {
+            _forward.Add(item1, item2);
+            _backward.Add(item2, item1);
```

Comment:
I'd still like to have an assert to highlight to anyone reading the code that there is this assumption that no added T1 ever equals an added T2.